### PR TITLE
Support cache auth with di

### PR DESCRIPTION
**Pull request by joaoantunes**
- fixed some compile issues (missing usings)
- consolidate the versions
- fixed the code regarding Authorize Caching (it was not working, because GetPolicy was not being able to provide the correct one, "cast" was failing)
- added a simple sample
- changed to choose the policy on the register by options (now uses DI) instead of a manual instance
- caching strategy is now choosen by the options

### CODE REVIEW

**Comment by yringler** [comment]
Nice! I'll try to look it over either today or tomorrow.
One thing - I use tabs for indentation, not spaces.
My apologies - this is the first time that a repo of mine has had someone else who wanted to contribute it, so I've never had to look in to writing style guides / contributing guides etc.

**Comment by yringler** [comment]
It seems that you didn't update the git submodule to the commit specified. This caused the build errors. (There is no Internal namespace on v3).
I fixed them, went back to v3, and pushed a commit.

See git submodule update and https://stackoverflow.com/a/49427199/2673256

**Comment by yringler** [comment]
After the dust settles with all this, will you consider taking ownership of this repository?
In the end, it will not be used where I work, and I think it would be healthiest if the maintainer is someone who's using it in production. Although I would probably add my name to the README 😉 .
In general, it's somewhat astonishing to me that @dotnet doesn't allow distributed response caching and caching secure responses by default. We're all professional programmers, we know what we're doing, and if @dotnet doesn't let we'll do it any way (ie, this project) 🤷 .
I would have to look in to how to transfer the nuget account to you, or add you as someone who can publish updates.

**Comment by joaoantunes** [comment]
Sorry I never used git submodules before. I've run git submodule update and got some errors:

image
Maybe my path to the repository must be shorter?

After that, I tried to compile the code but I have some errors:

image
Any idea what is causing this?

Don't I need to specify a version when making git submodule update?

**Comment by yringler** [comment]
>  Maybe my path to the repository must be shorter?

Yes. I put mine right at the root (C:\), and changed the name of the repo folder to cache, or something else short. Couldn't remember why I had put it  in such a random location! But yes, that was the reason.

> After that, I tried to compile the code but I have some errors:

Probably related to the submodule errors  - not all the code is there. Move the repo to a smaller path, run `git submodule update` again, clean solution, and try building again

> Don't I need to specify a version/tag when making git submodule update?

Git keeps track somewhere what commit is in the submodule. You can experiment out by checking out a different commit in the submodule, and then run `git diff` in the parent repo.

**Comment by joaoantunes** [comment]
Ok I've changed to a smaller repo.

The thing is when I clone I clone from master that uses the old module.
Then I accessed the aspnetcore submodule and made "git checkout tags/v3.1.6"
It retrieve properly the branch, but now I have this error when trying to compile:
image

Maybe I need to clean/compile first the aspnetcore using their tools? 

Does it work on your code?

**Comment by yringler** [comment]
> Then I accessed the aspnetcore submodule and made "git checkout tags/v3.1.6"

You shouldn't have to do that manually. Maybe try to re clone, take it from a clean slate.
1. Run `git config --global submodule.recurse true`. This will make sure that the submodules are always at the commit that the parent repo wants them to be at
1. Clone the repo
2. Make sure it's a small folder name
3. checkout the feature branch

> Does it work on your code?

I haven't had any issues building

Does you're version of visual studio [support](https://devblogs.microsoft.com/visualstudio/dot-net-core-support-in-visual-studio-2019-version-16-3/#:~:text=Visual%20Studio%2016.3%20supports%20both,back%2Dend%20microservices%20using%20gRPC.) netcore3?

> https://pkgs.dev.azure.com/dnceng/_packaging/darc-int-dotnet-extensions-be18161f

I don't recognize this azure URL

**Comment by joaoantunes** [comment]
My VS: Microsoft Visual Studio Enterprise 2019 **Version 16.6.4**

> You shouldn't have to do that manually. Maybe try to re clone, take it from a clean slate.
I did all the steps that you mention and still got the same errors when compiling the distributed-middleware-cache.sln

Can you try a clean clone to check if it works?

**Comment by yringler** [comment]
I also had issues with a clean clone.
I removed and re-added the submodule and pushed to this branch, and tried again.
So - in addition to `git submodule init`, also need to run ` git submodule update` to actually get the code.
IDK why the global config wasn't working.
Build succeeded.

**Comment by joaoantunes** [comment]
If I clone the last version I still got the same issue.
I think its because you always have to clone from the master, which has an older module. 

I posted this: https://github.com/dotnet/aspnetcore/issues/24013

I've tried to put the submodule to version to tag v3.1.0
Made a "dotnet restore" on the root, and it made the restore and I was able to compile!
But then I tried a second time to be sure from a clean clone I some other issues arise :(

**Comment by yringler** [comment]
I made a [gist](https://gist.github.com/yringler/d38332217a0a34b246f165e0d89e8d6b) with a bash script to
clone, checkout, and build.
I did see that running dotnet build causes that error - I don't know what that's all about, but building the projects (as in the gist) works.

I did see an error when I run the exe in the project though, but I don't think it's related.
```
Microsoft.AspNetCore.Hosting.Diagnostics[6]
      Application startup exception
System.InvalidOperationException: Unable to resolve service for type 'Microsoft.AspNetCore.ResponseCaching.IResponseCachingKeyProvider' while attempting to activate 'ExternalNetcoreExtensions.Custom.CustomResponseCachingMiddleware'.
   at Microsoft.Extensions.Internal.ActivatorUtilities.ConstructorMatcher.CreateInstance(IServiceProvider provider)
   at Microsoft.Extensions.Internal.ActivatorUtilities.CreateInstance(IServiceProvider provider, Type instanceType, Object[] parameters)
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass4_0.<UseMiddleware>b__0(RequestDelegate next)
   at Microsoft.AspNetCore.Builder.ApplicationBuilder.Build()
   at Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(CancellationToken cancellationToken)
Unhandled exception. System.InvalidOperationException: Unable to resolve service for type 'Microsoft.AspNetCore.ResponseCaching.IResponseCachingKeyProvider' while attempting to activate 'ExternalNetcoreExtensions.Custom.CustomResponseCachingMiddleware'.
   at Microsoft.Extensions.Internal.ActivatorUtilities.ConstructorMatcher.CreateInstance(IServiceProvider provider)
   at Microsoft.Extensions.Internal.ActivatorUtilities.CreateInstance(IServiceProvider provider, Type instanceType, Object[] parameters)
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass4_0.<UseMiddleware>b__0(RequestDelegate next)
   at Microsoft.AspNetCore.Builder.ApplicationBuilder.Build()
```

**Comment by joaoantunes** [comment]
> I did see an error when I run the exe in the project though, but I don't think it's related.

I know now with the internals the DI doesn't work, we need to provide the instance, I can fix that. 
Is not related with the compile issue. 

Any idea how to fix this? 
Check this issue that I've added: https://github.com/dotnet/aspnetcore/issues/24013#issuecomment-659575908

**Comment by yringler** [comment]
> Any idea how to fix those compile issues from VS?

What compile issues? I didn't see any build issues when I just build the distributedmiddlewarecache project and the example project.

> Check this issue that I've added: dotnet/aspnetcore#24013 (comment), regarding the packages from their private nugget repository.
They are saying if we remove "darc-int" from the NuGet.cofig it will work.

Neat. One benefit of relying on dotnet/aspnetcore is to give people a feeling of security. Caching can mean sensitive data, and people may want to know that it's a trusted entity whose code is handling the caching.
I think using a fork of dotnet/aspnetcore with a modified NuGet.config is acceptable... _if_ it's really needed.
What problem is there if we can't build the solution?
The purpose of this package is 2 things: the NuGet package, and the example, both of which build and run. Although I see the aesthetic benefit of having a build-able solution, and not just build-able projects, I don't think it's worth removing that direct dependency on dotnet/aspnetcore, especially if hopefully they'll fix that issue in an upcoming release.

**Comment by yringler** [comment]
Perhaps in the solution we can specify not to try building the aspnetcore code, only the nuget package and example?

**Comment by joaoantunes** [comment]
> What compile issues? I didn't see any build issues when I just build the distributedmiddlewarecache project and the example project.

Ok I got it what is the difference, I was compiling the solution, and It didn't work because was not able to compile the Microsoft.AspNetCore.ResponseCaching project.

> Perhaps in the solution we can specify not to try building the aspnetcore code, only the nuget package and example?
That would make sense then.

Since you are only using the file by hyperlink you don't use the .dlls/project right? 
That is the only way to bypass the internal issues with submodules right? The only alternative that I was seeing is to change the project of Microsoft.AspNetCore.ResponseCaching to make visible the Internals on distributedmiddlewarecache project, but since this implies changing an AspNetCore project, I think is only possible by a fork. 

> I did see an error when I run the exe in the project though, but I don't think it's related.
Tomorrow I'll push a fix

**Comment by yringler** [comment]
> That would make sense then.

To be clear, I'm talking about modifying DistributedResponseCachingMiddleware.sln. IDK how to do that, but it's probably possible.

> Since you are only using the file by hyperlink you don't use the .dlls/project right?

Correct

> That is the only way to bypass the internal issues with submodules right?

Yes, that is the only way to bypass the internal issue - because now, our code is on the inside, their code and ours are part of the same csproj. If you'll excuse the backpatting, I think it's quite clever.

> Tomorrow I'll push a fix

Great!

**Comment by joaoantunes** [comment]
> To be clear, I'm talking about modifying DistributedResponseCachingMiddleware.sln. IDK how to do that, but it's probably possible.

I've done it in this last commit. 

> Yes, that is the only way to bypass the internal issue - because now, our code is on the inside, their code and ours are part of the same csproj. If you'll excuse the backpatting, I think it's quite clever.

Yup that was a very nice idea! 💯 

With the v.3.1.6 because it had the internals I had to instantiate manually those who had only internal constructors, because DI container was not able to solve the dependencies. 
I've tested the code and seems to be working well 👍

File: DistributedResponseCachingMiddleware/Custom/CustomResponseCachingServiceExtensions.cs [hunk 1]
```diff
@@ -1,51 +1,78 @@
...
+    /// 
+    /// Note that this is a copypasta from <see cref="DistributedResponseCachingServicesExtensions"/>. I added it
+    /// to make the API symmetrical.
+    /// 
+    /// In a later version, perhaps DistributedResponseCaching will be deprecated, and the CustomResponseCache will use either
+    /// an IDistributedCache, or a ICustomResponseCache if it's available.
+    /// </remarks>
+    public static class CustomResponseCachingServiceExtensions
+    {
+        public static IServiceCollection AddDefaultResponseCaching(this IServiceCollection services)
```
Comment (yringler) [inline DistributedResponseCachingMiddleware/Custom/CustomResponseCachingServiceExtensions.cs:28]: Comment (docstring)

File: DistributedResponseCachingMiddleware/Custom/CustomResponseCachingServiceExtensions.cs [hunk 2]
```diff
@@ -1,51 +1,78 @@
...
+            };
+            if (options.CacheAuthorizedRequest)
+            {
+                services.AddCacheAuthorizedRequestsResponseCachingPolicy();
+            }
+            services.AddResponseCaching();
+            return services;
+        }
+
+        private static IServiceCollection RegisterResponseCache<T>(this IServiceCollection services, Func<IServiceProvider, T> responseCache = default)
```
Comment (yringler) [inline DistributedResponseCachingMiddleware/Custom/CustomResponseCachingServiceExtensions.cs:70]: Comment (docstring)

File: DistributedResponseCachingMiddleware/Utility/CustomResponseCachingPolicyProviderExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+﻿using Microsoft.AspNetCore.ResponseCaching;
+using Microsoft.Extensions.DependencyInjection;
+using System;
+
+namespace ExternalNetcoreExtensions.Utility
```
Comment (yringler) [inline DistributedResponseCachingMiddleware/Utility/CustomResponseCachingPolicyProviderExtensions.cs:5]: All extensions to IServiceCollection should be in the ` Microsoft.Extensions.DependencyInjection` namespace. This makes them more discoverable with intellisence.

File: DistributedResponseCachingMiddleware/Utility/CustomResponseCachingPolicyProviderExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+{
+    /// <summary>
+    /// Extension methods for the ResponseCaching Policies.
+    /// </summary>
+    public static class CustomResponseCachingPolicyProviderExtensions
+    {
+        /// <summary>
+        /// Add policy that allows caching even if it exists an Authorization header.
```
Comment (yringler) [inline DistributedResponseCachingMiddleware/Utility/CustomResponseCachingPolicyProviderExtensions.cs:13]: Grammar: even if an Authorization header exists
(This is more a note - I'll fix grammer things before I merge this in if you haven't already)

Comment (joaoantunes) [reply DistributedResponseCachingMiddleware/Utility/CustomResponseCachingPolicyProviderExtensions.cs]: I've made a commit  to solve the above issues, please check it.

**Comment by joaoantunes** [comment]
I was testing the Distributed caching implementation.
I've started the same API application two times, on port 5001 & 5002. 
I was expecting when asking on port 5001, if I asked for the same on 5002 the response would use the cached 5001 response, but it didn't work.

I've made some investigation and the issue was, the key that was being used to store on Redis had information about the port. 
This is because the implementation of IResponseCachingKeyProvider includes that information. 

`7/ GET<delimiter>SCHEME<delimiter>**HOST:PORT**/PATHBASE/PATH
public string CreateBaseKey(ResponseCachingContext context)
{
	if (context == null)
	{
		throw new ArgumentNullException(nameof(context));
	}

	var request = context.HttpContext.Request;
	var builder = _builderPool.Get();

	try
	{
		builder
			.AppendUpperInvariant(request.Method)
			.Append(KeyDelimiter)
			.AppendUpperInvariant(request.Scheme)
			.Append(KeyDelimiter)
			.AppendUpperInvariant(request.Host.Value);

		if (_options.UseCaseSensitivePaths)
		{
			builder
				.Append(request.PathBase.Value)
				.Append(request.Path.Value);
		}
		else
		{
			builder
				.AppendUpperInvariant(request.PathBase.Value)
				.AppendUpperInvariant(request.Path.Value);
		}

		return builder.ToString();
	}
	finally
	{
		_builderPool.Return(builder);
	}
}`

To solve this the only way that I've found was to make my own implementation of IResponseCachingKeyProvider 
Where is a copy-paste of ResponseCachingKeyProvider but removes the below lines:
```
.Append(KeyDelimiter)
.AppendUpperInvariant(request.Host.Value);
```

Do you think this approach makes sense?

**Comment by yringler** [comment]
It does make sense, but that's more copypasta than is needed.
Our options are limited because the base method isn't virtual, but we could create another implementation of `IResponseCachingKeyProvider` which is implemented through a `private readonly ResponseCachingKeyProvider`.
For example:
```c#
internal class CustomResponseCachingKeyProvider : IResponseCachingKeyProvider
{
  private readonly ResponseCachingKeyProvider baseKeyProvider;

  public (ObjectPoolProvider objectPoolProvider, IOptions<ResponseCachingOptions> options)
  {
    baseKeyProvider = new ResponseCachingKeyProvider(objectPoolProvider, options);
  }

  public string CreateBaseKey(ResponseCachingContext context)
  {
    string port = context.HttpContext.Request.Host.Value.ToUpperInvariant();
    return baseKeyProvider.CreateBaseKey(context).replace(port , "")
  }
}
```

And the other interface methods can just call the base method directly.
If you have the readonly property, VS will offer to implement the interface through it, and it'll generate the code for you.

I'm going to merge this PR, and hope to do some final things (grammer, update README.md) in the next few days, than I'll publish to NuGet.

**Comment by yringler** [comment]
Nice work, @joaoantunes !
I particularly liked the `AddResponseCaching` method. A very nice way to gather in one place all the functioanlity which had been kinda scattered around.
Also, the first time I've seen pattern matching used. Very cool.

**Comment by joaoantunes** [comment]
> Our options are limited because the base method isn't virtual, but we could create another implementation of IResponseCachingKeyProvider which is implemented through a private readonly ResponseCachingKeyProvider.

Yup... I'm only afraid about the "Replace" could hit more stuff and replace more than expected. Maybe we should replace only replace the first hit. I think we replace also the "host" not only the "port", not sure if the "scheme" should be replaced also. 

In my company, was decided that we can't depend on external packages like this one :(, so we are going to implement a similar approach that was developed here but in an internal repository. 
I have already a solution implemented on our repository for this situation. I can replicate it here if you want.

**Comment by yringler** [comment]
>  I can replicate it here if you want.

No, but thank you for offering.

> In my company, was decided that we can't depend on external packages like this one

An interesting policy, but to each their own I suppose.

> so we are going to implement a similar approach that was developed here

How similar, I wounder 😉 

Anyway, thank you for your contribution, it should be live on nuget later this week, and hopefully others will find it useful. All the best.

**Comment by joaoantunes** [comment]
> An interesting policy, but to each their own I suppose.

Unfortunately yes, is not open source, for now, maybe in the future they can change their mind.

> Anyway, thank you for your contribution, it should be live on nuget later this week, and hopefully others will find it useful. All the best.

Cool! If you need more help I'm glad to do it.

One thing that I did in that project is to have on the project just to reference the hyperlinks
And created another project for the implementation. Just to not mix the Microsoft code with the custom one.
To have access to the internals you just need to edit the ".csproj" where you added the Microsoft hyperlinks and add:

```
<ItemGroup>
		<AssemblyAttribute Include="System.Runtime.CompilerServices.InternalsVisibleTo">
			<_Parameter1>"PUT_HERE_THE_PROJECT_NAME_WHERE_YOU_WANT_TO_SEE_INTERNALS"</_Parameter1>
		</AssemblyAttribute>
	</ItemGroup>
```

Thanks for the help and good luck for this and future projects.

**Comment by yringler** [comment]
> One thing that I did in that project is to have on the project just to reference the hyperlinks
And created another project for the implementation. Just to not mix the Microsoft code with the custom one.
To have access to the internals you just need to edit the ".csproj" where you added the Microsoft hyperlinks and add:

That's a really great idea! Thanks for sharing.

**Pull request merged**